examples/src/custom_frames/fib1/air.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use super::{BaseElement, FibEvaluationFrame, FieldElement, ProofOptions, TRACE_WIDTH};
+use crate::utils::are_equal;
+use winterfell::{
+    Air, AirContext, Assertion, EvaluationFrame, TraceInfo, TransitionConstraintDegree,
+};
+
+// FIBONACCI AIR
+// ================================================================================================
+
+pub struct FibAir {
+    context: AirContext<BaseElement>,
+    result: BaseElement,
+}
+
+impl Air for FibAir {
+    type BaseField = BaseElement;
+    type PublicInputs = BaseElement;
+    type Frame<E: FieldElement> = FibEvaluationFrame<E>;
+    type AuxFrame<E: FieldElement> = FibEvaluationFrame<E>;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    fn new(trace_info: TraceInfo, pub_inputs: Self::BaseField, options: ProofOptions) -> Self {
+        let degrees = vec![TransitionConstraintDegree::new(1)];
+        assert_eq!(TRACE_WIDTH, trace_info.width());
+        let context =
+            AirContext::new(trace_info, degrees, 3, options).set_num_transition_exemptions(2);
```

Comment:
Maybe the addition of the exemptions shouldn't be done here nor through the Air context. By setting the number of exemptions you're basically modifying the divisor of `TrasitionConstraints` returned by `Air::get_transition_constraints`. Maybe the best that `get_transtition_constraints` is rewritten as follows
```
fn get_transition_constraints<E: FieldElement<BaseField = Self::BaseField>>(
        &self,
        composition_coefficients: &[(E, E)],
    ) -> TransitionConstraints<E> {
         // build constraint divisor; the same divisor applies to all transition constraints
        let divisor = ConstraintDivisor::from_transition(
            self.context().trace_len(),
            2,
        );
        TransitionConstraints::new(self.context(), divisor, composition_coefficients)
    }
```
Note that this also requires a small change on `TransitionConstraints::new` and in the default definition of  `get_transition_constraints`

Reply:
I forgot to mention that I also proposed to move the computation of the divisor because now it depends on `Self::Frame::FRAME_SHIFT`. The method `ConstraintDivisor::from_transition` should now receive the frame shift instead of the exemptions, and compute the divisor which vanishes over the set `{g^0, g^1,...,g^{(trace_lenght-frame_legth)/(frame_shift)}}` (I guess this should be also a sparse poly).

Reply:
I realized that a frame shift of 1 is actually sufficient for my purposes right now (implementing the Cairo constraints), so I went ahead and removed it from the PR with the intention of adding it back in at a later time. However, I do need to specify transition-specific divisors to make progress. I believe in my case that the number of exemptions will be a function of the public inputs. Maybe a vector of exemptions can be stored in `AirContext`, and `get_transition_constraints` can then use this to compute the appropriate divisors?

Reply:
I still need to write an additional test case, but I've pushed my changes if you want to take a look. I've added columns in the `ConstraintEvaluationTable` for each unique `TransitionConstraintGroup` (as determined by the tuple `(constraint_evaluation_degree, divisor_degree)`). As a consequence, there is now an extra column present in this table in the event that all constraint divisors are the same and any aux columns are present, as there is no merging between main and aux constraint groups. There is also a slight performance hit when merging constraints into a `TransitionConstraintGroup` in the case that all divisors are the same, as division of the divisor occurs at each step of the accumulator when folding, and not at the end. Both of these could be remedied, but I don't have a sense of whether the optimization is worth it right now.

Reply:
> I realized that a frame shift of 1 is actually sufficient for my purposes right now (implementing the Cairo constraints),

Without periodic values?

Reply:
and how many rows?

Reply:
Sorry! I wanted to sat columns! :smile: I was getting mad trying to figure out how could you do it with just two columns (but ~16 rows) and no frame_shifts.